File: code/dobot_quad_sdk-main/low_level/python/e2_depth_image_sub.py
import dds_middleware_python as dds
import sys
import time
import os
import cv2
import numpy as np


def depth_image_callback(depth_msg):
    """
    Callback for Image message when received
    """
    print(f"Received Image message:")
    print(
        f"  Timestamp: {depth_msg.header().stamp().sec()}.{depth_msg.header().stamp().nanosec():09d}"
    )
    print(f"  Frame ID: {depth_msg.header().frame_id()}")
    print(f"  Image size: {depth_msg.width()}x{depth_msg.height()}")
    print(f"  Encoding: {depth_msg.encoding()}")
    print(f"  Data size: {len(depth_msg.data())} bytes")
    print(f"  Step: {depth_msg.step()}")
    print(f"  Big endian: {depth_msg.is_bigendian()}")

    if "16UC1" in depth_msg.encoding():
        # Convert raw data to numpy array
        raw_data = np.array(depth_msg.data(), dtype=np.uint8)
        # Convert to 16-bit depth map
        depth_img = raw_data.view(np.uint16).reshape((depth_msg.height(), depth_msg.width()))
        # Normalize to 0-255
        depth_vis = cv2.normalize(depth_img, None, 0, 255, cv2.NORM_MINMAX, dtype=cv2.CV_8U)
        # Apply pseudo-color (Jet Colormap)
        depth_color = cv2.applyColorMap(depth_vis, cv2.COLORMAP_JET)
        filename = f"depth_images/depth_{depth_msg.header().stamp().sec()}_{depth_msg.header().stamp().nanosec()}.png"
        cv2.imwrite(filename, depth_color)
        print(f"Saved visibility depth map to {filename}")

    print("---")


def main():
    """
    Main function - DDS Image subscriber example
    """
    try:
        print("Starting DDS Python Image subscriber...")

        # Create save directory
        os.makedirs("depth_images", exist_ok=True)

        # Create DDS middleware instance with config file
        config_file = "config/dds_config.yaml"
        middleware = dds.PyDDSMiddleware(config_file)

        # Subscribe to Image topic
        topic_name = "rt/camera/camera2/image_depth"

        print(f"Subscribing to topic: {topic_name}")

        # Configure QoS parameters
        qos_config = {
            "reliability": "best_effort",
            "history_kind": "keep_last",
            "history_depth": 5,
            "durability": "volatile"
        }

        print(f"Using QoS config: {qos_config}")

        # Subscribe to Image topic and register callback with QoS
        middleware.subscribeImage(topic_name, depth_image_callback, qos_config)

        print("Image subscriber started, waiting for messages...")
        print("Press Ctrl+C to exit")

        # Keep program running to receive messages
        while True:
            time.sleep(1)

    except KeyboardInterrupt:
        print("\nReceived interrupt signal, exiting")
    except Exception as e:
        print(f"Error occurred: {e}")
        import traceback
        traceback.print_exc()
        return 1

    return 0


if __name__ == "__main__":
    sys.exit(main())

File: code/dobot_quad_sdk-main/low_level/cpp/e2_depth_image_sub.cc
#include <iostream>
#include <thread>
#include <chrono>
#include <cstdlib>
#include <opencv2/opencv.hpp>
#include "dds_middleware.hpp"
#include "sensor_msgs/msg/Image_.hpp"

using namespace dds_middleware;

void depth_callback(const sensor_msgs::msg::dds_::Image_& data)
{
    std::cout << "Received Depth Image:" << std::endl;
    std::cout << "sec=" << data.header_().stamp_().sec_() << std::endl;
    std::cout << "nanosec=" << data.header_().stamp_().nanosec_() << std::endl;
    std::cout << "frame_id=" << data.header_().frame_id_() << std::endl;
    std::cout << "height=" << data.height_() << std::endl;
    std::cout << "width=" << data.width_() << std::endl;
    std::cout << "encoding=" << data.encoding_() << std::endl;
    std::cout << "data_size=" << data.data_().size() << " bytes" << std::endl << std::endl;

    cv::Mat depth_img(data.height_(), data.width_(), CV_16UC1, const_cast<unsigned char*>(data.data_().data()));

    cv::Mat depth_vis;
    // Normalize 16-bit depth values to 0-255 range (CV_8UC1)
    cv::normalize(depth_img, depth_vis, 0, 255, cv::NORM_MINMAX, CV_8UC1);
    // Apply pseudo-color (Jet Colormap: Red/warm for near, Blue/cold for far)
    cv::Mat depth_color;
    cv::applyColorMap(depth_vis, depth_color, cv::COLORMAP_JET);

    std::string filename = "depth_images/depth_" + std::to_string(data.header_().stamp_().sec_()) + "_"
                           + std::to_string(data.header_().stamp_().nanosec_()) + ".png";

    cv::imwrite(filename, depth_color);
    std::cout << "Saved visibility depth map to " << filename << std::endl;
}

int main()
{
    system("mkdir -p depth_images");
    DDSMiddleware middleware("./config/dds_config.yaml");

    auto topic = middleware.createTopic<sensor_msgs::msg::dds_::Image_>("rt/camera/camera2/image_depth");

    auto reader = middleware.createReader<sensor_msgs::msg::dds_::Image_>(topic, depth_callback);

    std::cout << "Subscribed to depth image topic. Waiting for messages..." << std::endl;
    std::this_thread::sleep_for(std::chrono::hours(1));

    return 0;
}
